File: windows/src/CodexAyarTest.cs
namespace YzlabKurucu;

/// `CodexAyar` saf mantiginin testleri. macOS'taki `CodexAyarTest.swift` ile AYNI vakalar.
internal static class CodexAyarTest
{
    private const string SablonMetni =
        "# YapayZekaLab — Codex profili\n" +
        "model = \"gpt-5.6-sol\"\n" +
        "model_provider = \"yapayzekalab\"\n" +
        "model_catalog_json = \"C:\\\\Users\\\\x\\\\.codex\\\\yzlab-model-catalog.json\"\n" +
        "model_context_window = 450000\n" +
        "approval_policy = \"never\"\n" +
        "\n" +
        "[tools]\n" +
        "web_search = true\n" +
        "\n" +
        "[model_providers.yapayzekalab]\n" +
        "name = \"YapayZekaLab\"\n" +
        "base_url = \"https://yapayzekalab.org/v1\"\n" +
        "wire_api = \"responses\"\n" +
        "experimental_bearer_token = \"yzk_live_TEST\"\n";

    private const string O1 =
        "model = \"kendi-modelim\"\n" +
        "model_reasoning_effort = \"high\"\n" +
        "notify = [\"x\"]\n" +
        "\n" +
        "[mcp_servers.foo]\n" +
        "command = \"bar\"\n" +
        "\n" +
        "[model_providers.yapayzekalab]\n" +
        "name = \"eski\"\n" +
        "base_url = \"http://eski\"\n" +
        "\n" +
        "[model_providers.yapayzekalab.http_headers]\n" +
        "X = \"1\"\n" +
        "\n" +
        "[projects.\"/a\"]\n" +
        "trust_level = \"trusted\"\n";

    internal static List<string> Kume(string s) =>
        CodexAyar.SatirlaraAyir(s).Satirlar.Select(x => x.Trim()).Where(x => x.Length > 0).OrderBy(x => x, StringComparer.Ordinal).ToList();

    private static int Say(string metin, string parca) =>
        metin.Split(parca).Length - 1;

    public static void Calistir(Action<bool, string> kontrol)
    {
        var sb = CodexAyar.SablonuAyir(SablonMetni);
        kontrol(sb.Anahtarlar.SequenceEqual(new[] { "model", "model_provider", "model_catalog_json", "model_context_window", "approval_policy" }),
                "toml: sablon ust anahtarlari ayrildi");
        kontrol(sb.SaglayiciSatirlar.FirstOrDefault() == "[model_providers.yapayzekalab]" && sb.SaglayiciSatirlar.Count == 5,
                "toml: sablon saglayici tablosu ayrildi");
        kontrol(!sb.UstSatirlar.Any(x => x.Contains("web_search")) && !sb.SaglayiciSatirlar.Contains("[tools]"),
                "toml: [tools] tablosu ALINMADI (cakisma riski)");

        var (k1, y1) = CodexAyar.Uygula(O1, sb);
        kontrol(CodexAyar.Dogrula(k1) is null, "toml: kurulum sonrasi dosya gecerli (" + (CodexAyar.Dogrula(k1) ?? "ok") + ")");
        kontrol(Say(k1, "model_provider = \"yapayzekalab\"") == 1, "toml: model_provider tek sefer");
        kontrol(!k1.Contains("kendi-modelim") && y1.UstSatirlar.SequenceEqual(new[] { "model = \"kendi-modelim\"" }),
                "toml: musterinin model satiri yedege alindi");
        kontrol(k1.Contains("model_reasoning_effort = \"high\"") && k1.Contains("[mcp_servers.foo]") && k1.Contains("[projects.\"/a\"]"),
                "toml: musterinin diger ayarlari korundu");
        kontrol(!k1.Contains("http://eski") && !k1.Contains("http_headers"), "toml: eski saglayici tablolari kaldirildi");
        kontrol(y1.SaglayiciBlogu.Contains("[model_providers.yapayzekalab.http_headers]"), "toml: eski saglayici yedege alindi");
        kontrol(k1.IndexOf("model_provider = \"yapayzekalab\"", StringComparison.Ordinal) < k1.IndexOf("\n[", StringComparison.Ordinal),
                "toml: ust anahtarlar ilk tablodan ONCE");

        var (k2, _) = CodexAyar.Uygula(k1, sb);
        kontrol(k2 == k1, "toml: tekrar kurmak ayni dosyayi uretir (idempotent)");

        var g1 = CodexAyar.GeriAl(k1, y1);
        kontrol(g1 is not null && Kume(g1).SequenceEqual(Kume(O1)), "toml: geri al orijinal satirlari geri koydu");
        kontrol(g1 is not null && !g1.Contains("YapayZekaLab kurucu"), "toml: geri al isaretleri temizledi");

        var sonradan = k1 + "\n[projects.\"/b\"]\ntrust_level = \"trusted\"\n";
        var g2 = CodexAyar.GeriAl(sonradan, y1) ?? "";
        kontrol(g2.Contains("[projects.\"/b\"]") && Kume(g2).Count == Kume(O1).Count + 2,
                "toml: geri al sonradan eklenen tabloyu korudu");

        var (k3, y3) = CodexAyar.Uygula(null, sb);
        kontrol(CodexAyar.Dogrula(k3) is null && y3.DosyaYoktu, "toml: dosya yokken gecerli dosya uretildi");
        kontrol(CodexAyar.GeriAl(k3, y3) is null, "toml: dosya yoktuysa geri al dosyayi siler");

        var crlf = "\uFEFF" + O1.Replace("\n", "\r\n");
        var (k4, _) = CodexAyar.Uygula(crlf, sb);
        kontrol(CodexAyar.Dogrula(k4) is null && k4[0] != '\uFEFF' && Say(k4, "\n") == Say(k4, "\r\n"),
                "toml: CRLF korundu, BOM atildi");

        var bozuk = k1.Replace(CodexAyar.UstSon + "\n", CodexAyar.UstSon + "\nmodel = \"gpt-6-astra\"\n");
        kontrol(CodexAyar.Dogrula(bozuk) is not null, "toml: cift ust anahtar yakalandi");
        kontrol(CodexAyar.Dogrula(CodexAyar.Uygula(bozuk, sb).Metin) is null, "toml: yeniden kur cift anahtari temizledi");

        var (k5, _) = CodexAyar.Uygula("[mcp_servers.x]\ncommand = \"y\"\n", sb);
        kontrol(CodexAyar.Dogrula(k5) is null && k5.StartsWith(CodexAyar.UstBas), "toml: tablo ile baslayan dosya");
    }
}

File: macos/Sources/YzlabKurucu/CodexAyarTest.swift
import Foundation

/// `CodexAyar` saf mantığının testleri. Windows'taki `CodexAyarTest.cs` ile AYNI vakalar.
enum CodexAyarTest {
    static let sablonMetni = """
    # YapayZekaLab — Codex profili
    model = "gpt-5.6-sol"
    model_provider = "yapayzekalab"
    model_catalog_json = "/x/yzlab-model-catalog.json"
    model_context_window = 450000
    approval_policy = "never"

    [tools]
    web_search = true

    [model_providers.yapayzekalab]
    name = "YapayZekaLab"
    base_url = "https://yapayzekalab.org/v1"
    wire_api = "responses"
    experimental_bearer_token = "yzk_live_TEST"

    """

    static let o1 = """
    model = "kendi-modelim"
    model_reasoning_effort = "high"
    notify = ["x"]

    [mcp_servers.foo]
    command = "bar"

    [model_providers.yapayzekalab]
    name = "eski"
    base_url = "http://eski"

    [model_providers.yapayzekalab.http_headers]
    X = "1"

    [projects."/a"]
    trust_level = "trusted"

    """

    static func kume(_ s: String) -> [String] {
        CodexAyar.satirlaraAyir(s).satirlar
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
            .sorted()
    }

    static func say(_ metin: String, _ parca: String) -> Int {
        metin.components(separatedBy: parca).count - 1
    }

    static func calistir(_ kontrol: (Bool, String) -> Void) {
        let sb = CodexAyar.sablonuAyir(sablonMetni)
        kontrol(sb.anahtarlar == ["model", "model_provider", "model_catalog_json", "model_context_window", "approval_policy"],
                "toml: sablon ust anahtarlari ayrildi")
        kontrol(sb.saglayiciSatirlar.first == "[model_providers.yapayzekalab]" && sb.saglayiciSatirlar.count == 5,
                "toml: sablon saglayici tablosu ayrildi")
        kontrol(!sb.ustSatirlar.contains { $0.contains("web_search") } && !sb.saglayiciSatirlar.contains("[tools]"),
                "toml: [tools] tablosu ALINMADI (cakisma riski)")

        // Mevcut dosya: eski saglayici + alt tablo + musteri ayarlari
        let (k1, y1) = CodexAyar.uygula(o1, sb)
        kontrol(CodexAyar.dogrula(k1) == nil, "toml: kurulum sonrasi dosya gecerli (\(CodexAyar.dogrula(k1) ?? "ok"))")
        kontrol(say(k1, "model_provider = \"yapayzekalab\"") == 1, "toml: model_provider tek sefer")
        kontrol(!k1.contains("kendi-modelim") && y1.ustSatirlar == ["model = \"kendi-modelim\""],
                "toml: musterinin model satiri yedege alindi")
        kontrol(k1.contains("model_reasoning_effort = \"high\"") && k1.contains("[mcp_servers.foo]") && k1.contains("[projects.\"/a\"]"),
                "toml: musterinin diger ayarlari korundu")
        kontrol(!k1.contains("http://eski") && !k1.contains("http_headers"), "toml: eski saglayici tablolari kaldirildi")
        kontrol(y1.saglayiciBlogu.contains("[model_providers.yapayzekalab.http_headers]"), "toml: eski saglayici yedege alindi")
        let ilkBaslik = k1.range(of: "\n[")!.lowerBound
        kontrol(k1.range(of: "model_provider = \"yapayzekalab\"")!.lowerBound < ilkBaslik, "toml: ust anahtarlar ilk tablodan ONCE")

        // Tekrar kurulum: ayni sonuc, cift kayit yok
        let (k2, _) = CodexAyar.uygula(k1, sb)
        kontrol(k2 == k1, "toml: tekrar kurmak ayni dosyayi uretir (idempotent)")

        // Geri al: orijinal satirlarin hepsi geri gelir
        let g1 = CodexAyar.geriAl(k1, y1)
        kontrol(g1 != nil && kume(g1!) == kume(o1), "toml: geri al orijinal satirlari geri koydu")
        kontrol(g1.map { !$0.contains("YapayZekaLab kurucu") } ?? false, "toml: geri al isaretleri temizledi")

        // Kurulumdan sonra Codex'in ekledigi tablo geri almada korunur
        let sonradan = k1 + "\n[projects.\"/b\"]\ntrust_level = \"trusted\"\n"
        let g2 = CodexAyar.geriAl(sonradan, y1) ?? ""
        kontrol(g2.contains("[projects.\"/b\"]") && kume(g2).count == kume(o1).count + 2,
                "toml: geri al sonradan eklenen tabloyu korudu")

        // Dosya yoktu
        let (k3, y3) = CodexAyar.uygula(nil, sb)
        kontrol(CodexAyar.dogrula(k3) == nil && y3.dosyaYoktu, "toml: dosya yokken gecerli dosya uretildi")
        kontrol(CodexAyar.geriAl(k3, y3) == nil, "toml: dosya yoktuysa geri al dosyayi siler")

        // CRLF + BOM (Windows'ta yazilmis dosya)
        let crlf = "\u{FEFF}" + o1.replacingOccurrences(of: "\n", with: "\r\n")
        let (k4, _) = CodexAyar.uygula(crlf, sb)
        kontrol(CodexAyar.dogrula(k4) == nil && !k4.hasPrefix("\u{FEFF}") && say(k4, "\n") == say(k4, "\r\n"),
                "toml: CRLF korundu, BOM atildi")

        // Uygulama sonradan isaret DISINA ayni anahtari yazarsa: dogrula yakalar, yeniden kur duzeltir
        let bozuk = k1.replacingOccurrences(of: CodexAyar.ustSon + "\n", with: CodexAyar.ustSon + "\nmodel = \"gpt-6-astra\"\n")
        kontrol(CodexAyar.dogrula(bozuk) != nil, "toml: cift ust anahtar yakalandi")
        kontrol(CodexAyar.dogrula(CodexAyar.uygula(bozuk, sb).metin) == nil, "toml: yeniden kur cift anahtari temizledi")

        // Ilk satiri tablo olan dosya
        let (k5, _) = CodexAyar.uygula("[mcp_servers.x]\ncommand = \"y\"\n", sb)
        kontrol(CodexAyar.dogrula(k5) == nil && k5.hasPrefix(CodexAyar.ustBas), "toml: tablo ile baslayan dosya")
    }
}
